applications/asset_tracker/src/env_sensors/bsec.c

```diff
@@ -0,0 +1,271 @@
+			settings_read_cb read_cb, void *cb_arg)
+{
+
+	if (!strcmp(key, "state")) {
+		s_state_buffer_len = len_rd;
+		if (read_cb(cb_arg, s_state_buffer, len_rd) > 0) {
+			return 0;
+		}
+	}
+	s_state_buffer_len = 0;
+	return -1;
+}
+
+static int enable_settings(void)
+{
+	int err = 0;
+
+	settings_subsys_init();
+	struct settings_handler my_conf = {
+		.name = "bsec",
+		.h_set = settings_set
+	};
+
+	err = settings_register(&my_conf);
+	if (err) {
+		printk("Cannot register settings handler");
+		goto error;
+	}
+
+	/* This module loads settings for all application modules */
+	err = settings_load();
+	if (err) {
+		printk("Cannot load settings");
+		goto error;
+	}
+
+error:
+	return err;
+}
+
+static int8_t bus_write(u8_t dev_addr, u8_t reg_addr,
+			u8_t *reg_data_ptr, u16_t data_len)
+{
+	u8_t buf[data_len+1];
+
+	buf[0] = reg_addr;
+	memcpy(&buf[1], reg_data_ptr, data_len);
+	return i2c_write(i2c_master, buf, data_len+1, dev_addr);
+}
+
+static s8_t bus_read(u8_t dev_addr, u8_t reg_addr,
+			u8_t *reg_data_ptr, u16_t data_len)
+{
+	return i2c_write_read(i2c_master, dev_addr, &reg_addr,
+				1, reg_data_ptr, data_len);
+}
+
+static void sleep(u32_t t_ms)
+{
+	k_sleep(t_ms);
+}
+
+static s64_t get_timestamp_us(void)
+{
+	return k_uptime_get()*1000;
+}
+
+static void output_ready(s64_t timestamp, float iaq, u8_t iaq_accuracy,
+			float temperature, float humidity, float pressure,
+			float raw_temperature, float raw_humidity, float gas,
+			bsec_library_return_t bsec_status, float static_iaq,
+			float co2_equivalent, float breath_voc_equivalent)
+{
+	k_spinlock_key_t key = k_spin_lock(&(temp_sensor.lock));
+
+	temp_sensor.sensor.value = temperature;
+	k_spin_unlock(&(temp_sensor.lock), key);
+	key = k_spin_lock(&(humid_sensor.lock));
+	humid_sensor.sensor.value = humidity;
+	k_spin_unlock(&(humid_sensor.lock), key);
+	key = k_spin_lock(&(pressure_sensor.lock));
+	pressure_sensor.sensor.value = pressure/1000;
+	k_spin_unlock(&(pressure_sensor.lock), key);
+	key = k_spin_lock(&(air_quality_sensor.lock));
+	air_quality_sensor.sensor.value = iaq;
+	k_spin_unlock(&(air_quality_sensor.lock), key);
+}
+
+static u32_t state_load(u8_t *state_buffer, u32_t n_buffer)
+{
+	if ((s_state_buffer_len > 0) && (s_state_buffer_len <= n_buffer)) {
+		memcpy(state_buffer, s_state_buffer, s_state_buffer_len);
+		return s_state_buffer_len;
+	} else {
+		return 0;
+	}
+}
+
+static void state_save(const u8_t *state_buffer, u32_t length)
+{
+	settings_save_one("bsec/state", state_buffer, length);
+}
+
+static u32_t config_load(u8_t *config_buffer, u32_t n_buffer)
+{
+	/* Not implemented */
+	return 0;
+}
+
+static void bsec_thread(void)
+{
+	bsec_iot_loop(sleep, get_timestamp_us, output_ready, state_save, 12);
+}
+
+int env_sensors_get_temperature(env_sensor_data_t *sensor_data)
+{
+	if (sensor_data == NULL) {
+		return -1;
+	}
+	k_spinlock_key_t key = k_spin_lock(&temp_sensor.lock);
+
+	memcpy(sensor_data, &(temp_sensor.sensor), sizeof(temp_sensor.sensor));
```

Comment:
Why the parenthesis around `&(temp_sensor.sensor)` (here and below)?

Reply:
removed paranthesis